server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -1587,6 +1590,44 @@ public void testForceMergeWithSoftDeletesRetention() throws Exception {
             engine.forceMerge(true, 1, false, false, false);
             assertConsistentHistoryBetweenTranslogAndLuceneIndex(engine, mapperService);
             assertThat(readAllOperationsInLucene(engine, mapperService), hasSize(liveDocs.size()));
+            assertPruneIdBelowLocalCheckpointSafeCommit(engine, mapperService, safeCommitCheckpoint);
+        }
+    }
+
+    private void assertPruneIdBelowLocalCheckpointSafeCommit(Engine engine, MapperService mapperService,
+                                                             long localCheckpointOfSafeCommit) throws IOException {
+        try (Searcher search = engine.acquireSearcher("test", Engine.SearcherScope.INTERNAL)) {
+            for (LeafReaderContext leaf : search.getDirectoryReader().leaves()) {
+                LeafReader leafReader = leaf.reader();
+                Map<String, Set<Integer>> retainedSegmentDocIds = new HashMap<>();
+                NumericDocValues seqNoDV = leafReader.getNumericDocValues(SeqNoFieldMapper.NAME);
+                for (int segmentDocID = 0; segmentDocID < leafReader.maxDoc(); segmentDocID++) {
+                    seqNoDV.advanceExact(segmentDocID);
+                    if (seqNoDV.longValue() > localCheckpointOfSafeCommit || leafReader.getLiveDocs().get(segmentDocID)) {
+                        final FieldsVisitor fields = new FieldsVisitor(true, SourceFieldMapper.NAME);
+                        leafReader.document(segmentDocID, fields);
+                        fields.postProcess(mapperService);
+                        retainedSegmentDocIds.computeIfAbsent(fields.uid().id(), k -> new HashSet<>()).add(segmentDocID);
+                    }
+                }
+                TermsEnum termsEnum = leafReader.terms(IdFieldMapper.NAME).iterator();
+                if (termsEnum == null) {
+                    assertThat(retainedSegmentDocIds.keySet(), empty());
+                    continue;
+                }
+                for (String id : retainedSegmentDocIds.keySet()) {
+                    if (termsEnum.seekExact(newUid(id).bytes()) == false) {
+                        assertThat(retainedSegmentDocIds.get(id), empty());
```

Comment:
The way retainedSegmentDocIds is maintained, this should never be reached (since we always add to the set immediately after adding into the map in line 1610).

It would be nice to pass in all ids generated by the test into this method and prepopulate the retainedSegmentDocIds with empty sets for all of them. That would check all ids, including checking that deleted docs with max(seqNo) <= localCheckpointOfSafeCommit have no postings.